lib/analysis/modules/ies_enkf.c

```diff
@@ -0,0 +1,1059 @@
+#define IES_DEBUG_KEY                    "IES_DEBUG"
+
+//#define DEFAULT_ANALYSIS_SCALE_DATA true
+
+
+//**********************************************
+// IES "object" data definition
+//**********************************************
+/*
+  The configuration data used by the ies_enkf module is contained in a
+  ies_enkf_data_struct instance. The data type used for the ies_enkf
+  module is quite simple; with only a few scalar variables, but there
+  are essentially no limits to what you can pack into such a datatype.
+
+  All the functions in the module have a void pointer as the first
+  argument, this will immediately be casted to an ies_enkf_data_type
+  instance, to get some type safety the UTIL_TYPE_ID system should be
+  used.
+
+  The data structure holding the data for your analysis module should
+  be created and initialized by a constructor, which should be
+  registered with the '.alloc' element of the analysis table; in the
+  same manner the desctruction of this data should be handled by a
+  destructor or free() function registered with the .freef field of
+  the analysis table.
+*/
+
+struct ies_enkf_data_struct {
+   UTIL_TYPE_ID_DECLARATION;
+   int       iteration_nr;            // Keep track of the outer iteration loop
+   double    truncation;              // Controlled by config key: ENKF_TRUNCATION_KEY
+   int       subspace_dimension;      // Controlled by config key: ENKF_SUBSPACE_DIMENSION_KEY (-1: use Truncation instead)
+   double    ies_steplength;          // Step length in Gauss Newton iteration
+   double    gauss_newton_conv;       // NOT USED
+   bool      ies_subspace;            // NOT USED
+   int       ies_inversion;                // 1 for exact R, 2 for R=EE', 3 for E
+   char    * ies_logfile;             // Filename for ies logfile
+   bool      ies_debug;               // True or false for storing debug information
+   int       max_gauss_newton_it;     // NOT USED
+   int state_size;                    // Initial state_size used for checks in subsequent calls
+   bool_vector_type * ens_mask0;      // Initial ensemble mask of active realizations
+   bool_vector_type * ens_mask;       // Ensemble mask of active realizations
+   bool_vector_type * obs_mask0;      // Initial observation mask for active measurements
+   bool_vector_type * obs_mask;       // Current observation mask
+   matrix_type * W;                   // Coefficient matrix used to compute Omega = I + W (I -11'/N)/sqrt(N-1)
+   matrix_type * A0;                  // Prior ensemble used in Ei=A0 Omega_i
+   matrix_type * E;                   // Prior ensemble of measurement perturations (should be the same for all iterations)
+   bool      converged;               // GN has converged
+   ies_enkf_config_type * config;     // This I don't understand but I assume I include data from the ies_enkf_config_type defined in ies_enkf_config.c
+   bool      AAprojection;            // For including the AAprojection of Y
+};
+
```

Comment:
It seems that the division between the `ies_enkf_config` object and the `enkf_ies_data` struct is not really clear. There are of course no hard rules here - but an approach which I think would make sense (and I *think* that is how it is in the rml module):

**ies_config**: This object is for various settings which can be configured/set by the user. The values in the `ies_config`object should *not* be modified by the algorithm.

**ies_data**: This is (temporary) data needed by the algorithm while running.